File: framework/languages/dotnet/src/Zlink.Framework/Runtime/Locations/ZLinkLocationReadiness.cs
namespace Zlink.Framework.Runtime.Locations;

internal sealed class ZLinkLocationReadiness(IZLinkLocationRuntimeQuery query) : IZLinkLocationReadiness
{
    public async ValueTask<bool> IsPeerReadyAsync(
        string meshName,
        ZLinkLocationRole role,
        RoutingId? nodeRid = null,
        CancellationToken cancellationToken = default)
    {
        try
        {
            var page = await query.ListTopologyAsync(
                    new ZLinkLocationTopologyFilter(
                        MeshName: meshName,
                        NodeRid: nodeRid,
                        State: ZLinkLocationTopologyState.Ready),
                    cancellationToken: cancellationToken)
                .ConfigureAwait(false);
            return page.Items.Count > 0;
        }
        catch
        {
            return false;
        }
    }
}

File: framework/languages/java/zlink-framework-core/src/main/java/systems/zlink/framework/locations/ZLinkLocationReadiness.java
package systems.zlink.framework.locations;

import systems.zlink.contracts.core.RoutingId;
import systems.zlink.framework.runtime.internal.locations.*;

import java.util.concurrent.CompletionStage;

public interface ZLinkLocationReadiness {
    CompletionStage<Boolean> isPeerReady(
            String meshName, ZLinkLocationRole role, RoutingId nodeRid);
}
